### DISCUSSION TITLE
Can i do chart using sparkline?

### DISCUSSION THREAD

**Discussion by wstr27** [Textualize/textual#4503] [Q&A] [answered]
I want add chart in my programm, but i don't know how i can do it. I try to using sparklines, but i  don't have result that satisfied me

**Comment by davep** [comment, marked as answer]
`Sparkline` is, as the name suggests, for making sparklines. You say you didn't find it satisfying, but without saying what was dissatisfying, or what you would find satisfying, it's hard for anyone to really say.

That said... have you looked at [textual-plotext](https://github.com/Textualize/textual-plotext)?

**Reply by wstr27** [reply]
Thank you very much, i don't know about this and i can't find. This is exactly what I need!!!!

**Reply by wstr27** [reply]
If you are familiar with this topic, could you help with one problem? Why is my schedule not being drawn? Moreover, the sec, fps and fps data are available, I checked.
![Uploading Снимок экрана 2024-05-11 в 03.16.22.png…]()

**Reply by davep** [reply]
Without an MRE, or anything to show what you're doing or seeing, it's impossible to say. 

I'd encourage hopping into the discussions in the relevant repository, or even asking on the Textual Discord server, but please give full details of what you're doing, ideally as an MRE.

**Reply by wstr27** [reply]
Sorry, github uploading my pic

**Reply by wstr27** [reply]
If u can, we can go to telegram

**Reply by davep** [reply]
No idea what that is. The resources I mention are the best to get help via.

**Reply by wstr27** [reply]
`import gevent
from gevent import monkey
monkey.patch_all()
import locust
from textual.app import App, ComposeResult
from textual.containers import ScrollableContainer
from textual.widgets import Button, Footer, Header, Input, Pretty, Static, TabbedContent, TabPane, LoadingIndicator
from textual.binding import Binding
from textual import on
import speedtest
import time
import requests
from locust import HttpUser, events, task
from locust.env import Environment
from locust.log import setup_logging
from locust.stats import stats_history, stats_printer
import math
from textual_plotext import PlotextPlot

locust.stats.HISTORY_STATS_INTERVAL_SEC = 1

class MyUser(HttpUser):
    host="https://wstr27.ru"

    @task
    def test(self):
        self.client.get("/")


class MyApp(App):
    CSS_PATH = "main.tcss"
    BINDINGS = [
            Binding(key="?", action="quit", description="Выход"),
            Binding(key="q", action="show_tab('test1')", description="Динамические характеристики"),
            Binding(key="w", action="show_tab('test2')", description="Нагрузочное тестирование"),
            Binding(key="e", action="show_tab('test3')", description="Проверка интернет-соединения"),
            Binding(key="r", action="show_tab('test4')", description="Средства реализации сайта"),
        ]
    def compose(self) -> ComposeResult:
        self.title = "Супер программа"
        self.sub_title = " Гришанова Д.А."
        yield Header()
        yield Footer()
        with TabbedContent(initial='test1'):
            with TabPane('Динамические характеристики', id='test1'):
                yield Input(placeholder='url', id="input1")
                yield Button("Пуск", id="test1_button")
            with TabPane('Нагрузочное тестирование', id='test2'):
                yield Input(placeholder='url', id="input2_1")
                yield Input(placeholder='users', id="input2_2")
                yield Input(placeholder='spawn rate', id="input2_3")
                yield Input(placeholder='time', id="input2_4")
                yield Button("Пуск", id="test2_button")
            with TabPane("Проверка интернет-соединения", id="test3"):
                yield Button("Пуск", id="test3_button")
            with TabPane("Средства реализации сайта", id="test4"):
                yield Input(placeholder='url', id="input4")
                yield Button("Пуск", id="test4_button")


    def action_show_tab(self, tab: str) -> None:
            """Switch to a new tab."""
            self.get_child_by_type(TabbedContent).active = tab

    @on(Button.Pressed, "#test1_button")
    def button1_press(self):
        url = "https://www.googleapis.com/pagespeedonline/v5/runPagespeed"
        querystring = {"url": self.query_one("#input1").value,"key":"xxx"}
        response = requests.request("GET", url, params=querystring)
        answer = ''
        self.query_one("#test1").remove_children()
        self.query_one("#test1").mount(Input(placeholder='url', id="input1"))
        self.query_one("#test1").mount(Button("Пуск", id="test1_button"))
        if "error" in response.json():
            answer = f"Error with description:\n"
            for error in response.json()['error']['errors']:
                answer += error['message'] + '\n'
            self.query_one("#test1").mount(Static(answer))
        else:
            metrics = response.json()['lighthouseResult']['audits']['metrics']['details']['items'][0]
            self.query_one("#test1").mount(Static(f"url: {response.json()['id']}"))
            self.query_one("#test1").mount(Static(f"FCP: {metrics['firstContentfulPaint']}мс"))
            self.query_one("#test1").mount(Static(f"TBT: {metrics['totalBlockingTime']}мс"))
            self.query_one("#test1").mount(Static(f"LCP: {metrics['largestContentfulPaint']}мс"))
            self.query_one("#test1").mount(Static(f"TTI: {metrics['interactive']}мс"))
            self.query_one("#test1").mount(Static(f"CLS: {metrics['cumulativeLayoutShift']}"))
            self.query_one("#test1").mount(Static(f"TTFB: {metrics['timeToFirstByte']}"))
            self.query_one("#test1").mount(Static(f"Тут должен быть график"))

    @on(Button.Pressed, "#test2_button")
    def button2_press(self):
        need_url2 = self.query_one("#input2_1").value
        need_users = int(self.query_one("#input2_2").value)
        need_spawn_rate = int(self.query_one("#input2_3").value)
        need_time = int(self.query_one("#input2_4").value) + 1
        self.query_one("#test2").remove_children()
        self.query_one("#test2").mount(Input(placeholder='url', id="input2_1"))
        self.query_one("#test2").mount(Input(placeholder='users', id="input2_2"))
        self.query_one("#test2").mount(Input(placeholder='spawn rate', id="input2_3"))
        self.query_one("#test2").mount(Input(placeholder='time', id="input2_4"))
        self.query_one("#test2").mount(Button("Пуск", id="test2_button"))
        MyUser.host = need_url2
        env = Environment(user_classes=[MyUser])
        env.create_local_runner()
        gevent.spawn(stats_printer(env.stats))
        gevent.spawn(stats_history, env.runner)
        env.runner.start(need_users, spawn_rate=need_spawn_rate)
        time.sleep(need_time)
        env.runner.stop()
        env.runner.quit()
        sec = []
        rps = []
        fps = []
        avg = []
        users = []
        for i in range(len(env.stats.history)):
            sec.append(i)
            rps.append(round(env.stats.history[i]["current_rps"]))
            fps.append(round(env.stats.history[i]["current_fail_per_sec"]))
            avg.append(round(env.stats.history[i]["total_avg_response_time"]))
            users.append(round(env.stats.history[i]["user_count"]))
        self.query_one("#test2").mount(Static(f"reqs: {env.stats.num_requests}"))
        self.query_one("#test2").mount(Static(f"fails: {env.stats.num_failures}"))
        self.query_one("#test2").mount(Static(f"none_resps: {env.stats.num_none_requests}"))
        self.query_one("#test2").mount(Static(f"avg: {round(env.stats.get('/', 'GET').avg_response_time)}мс"))
        self.query_one("#test2").mount(Static(f"min: {env.stats.get('/', 'GET').min_response_time}мс"))
        self.query_one("#test2").mount(Static(f"max: {round(env.stats.get('/', 'GET').max_response_time)}мс"))
        self.query_one("#test2").mount(Static(f"median: {round(env.stats.get('/', 'GET').median_response_time)}мс"))
        self.query_one("#test2").mount(Static(f"req/s: {round(env.stats.get('/', 'GET').total_rps)}"))
        self.query_one("#test2").mount(Static(f"fails/s: {round(env.stats.get('/', 'GET').total_fail_per_sec)}"))
        self.query_one("#test2").mount(PlotextPlot(id="chart1"))
        plt = self.query_one("#chart1").plt
        plt.plot(sec, rps, marker="braille")
        plt.plot(sec,fps, marker="braille")
        plt.title("Requests/s and fails/s")
        plt.ylabel("requests")
        plt.xlabel("seconds")
        plt.grid(1, 1)
        plt.xfrequency(11)
        # self.query_one("#chart1").refresh()




    @on(Button.Pressed, "#test3_button")
    def button3_press(self):
        st = speedtest.Speedtest(secure=True)
        st.get_best_server()
        st.download()
        st.upload()
        st_result = st.results.dict()
        self.query_one("#test3").remove_children()
        self.query_one("#test3").mount(Static(f"Скорость скачивания: {round(st_result["download"])/1024/1024} Мбит/с"))
        self.query_one("#test3").mount(Static(f"Скорость загрузки: {round(st_result["upload"])/1024/1024} Мбит/с"))
        self.query_one("#test3").mount(Static(f"Пинг: {st_result["ping"]} мс"))
        self.query_one("#test3").mount(Button("Пройти ещё раз!", id="test3_button"))

    @on(Button.Pressed, "#test4_button")
    def button4_press(self):
        need_url = self.query_one("#input4").value
        self.query_one("#test4").remove_children()
        self.query_one("#test4").mount(Input(placeholder='url', id="input4"))
        self.query_one("#test4").mount(Button("Пуск", id="test4_button"))
        url = "https://api.builtwith.com/v21/api.json"
        querystring = {"KEY":"xxx","LOOKUP": need_url}
        response = requests.request("GET", url, params=querystring)
        if len(response.json()["Errors"]) > 0:
            self.query_one("#test4").mount(Static(f"{response.text}"))
        else:
            technologies = response.json()["Results"][0]["Result"]["Paths"][0]["Technologies"]
            for technology in technologies:
                self.query_one("#test4").mount(Static(f"{technology['Name']}: {technology['Description']}"))

if __name__ == "__main__":
    app = MyApp()
    app.run()
`

This my code.
It don't work at this moment:
`self.query_one("#test2").mount(PlotextPlot(id="chart1"))
        plt = self.query_one("#chart1").plt
        plt.plot(sec, rps, marker="braille")
        plt.plot(sec,fps, marker="braille")
        plt.title("Requests/s and fails/s")
        plt.ylabel("requests")
        plt.xlabel("seconds")
        plt.grid(1, 1)
        plt.xfrequency(11)`

I don't know why

**Reply by davep** [reply]
That's rather a lot of code to wade through and would seem to need a number of dependencies that people won't have.

As I have suggested a couple of times now: perhaps you could create [an MRE](https://github.com/Textualize/textual/discussions/3361) of what you're trying to do and then seek help either [in the discussions area of the textual-plotext repo](https://github.com/Textualize/textual-plotext/discussions), or if it's your preference, the `#help-wanted` channel on the [the Textual Discord](https://discord.gg/Enf6Z3qhVr).
